src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -309,94 +268,56 @@ private static ConstantValue TryFoldTwoConcatConsts(ConstantValue leftConst, Con
         /// </summary>
         private BoundExpression RewriteStringConcatenationOneExpr(SyntaxNode syntax, BoundExpression loweredOperand)
         {
-            if (loweredOperand.Type.SpecialType == SpecialType.System_String)
+            // If it's a call to 'string.Concat' (or is something which ends in '?? ""', which this method also extracts,
+            // we know the result cannot be null. Otherwise return loweredOperand ?? ""
+            if (TryExtractStringConcatArgs(loweredOperand, out _))
             {
-                // If it's a call to 'string.Concat(object) ?? ""' or another overload of 'string.Concat', we know the result cannot
-                // be null. Otherwise if it's 'string.Concat(object)' or something which isn't 'string.Concat', return loweredOperand ?? ""
-                if (TryExtractStringConcatArgs(loweredOperand, out _, out bool loweredCanReturnNull) && !loweredCanReturnNull)
-                {
-                    return loweredOperand;
-                }
-                else
-                {
-                    return _factory.Coalesce(loweredOperand, _factory.Literal(""));
-                }
+                return loweredOperand;
             }
+            else
+            {
+                return _factory.Coalesce(loweredOperand, _factory.Literal(""));
+            }
+        }
 
-            var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatObject);
+        private BoundExpression RewriteStringConcatenationTwoExprs(SyntaxNode syntax, ImmutableArray<BoundExpression> loweredArgs)
```

Comment:
I changed from taking individual arguments to taking an `ImmutableArray` simply to save a little cost and a little code. Happy to change it back.

Reply:
Seems like in the original we just wound up creating an ImmutableArray a little bit later anyway. So I don't see that much advantage to the old way.

Reply:
Yeah there's not much in it - made is slightly easier to assert that all were strings, and avoided the boilerplate of splitting them out into locals and then back again. I'll put it back.